app/api/resources/metadata.py

```diff
@@ -44,5 +45,17 @@ async def get_metadata_statistics(request: Request, genome_uuid: str):
         genome_stats = GenomeStatistics(_raw_data=top_level_stats_dict["statistics"])
         return responses.JSONResponse({"genome_stats": genome_stats.dict()})
     except Exception as e:
-        logger.log("INFO", e)
+        return response_error_handler({"status": 500})
+
+@router.get("/popular_species", name="popular_species")
+async def get_popular_species(request: Request):
+    try:
+        popular_species_dict = MessageToDict(grpc_client.get_popular_species())
+        popular_species = popular_species_dict['organismsGroupCount']
+        for species in popular_species:
```

Comment:
I'm wondering if this part can be done in models while the objects are being created using a Pydantic @validator .

Reply:
I wish I could but in order to get the base_url we have to modify it here or else we will have to pass the request object or base_url with all the request. Which I think is overkill.

Reply:
It would be better to follow the MVC design patter and keep data logic in the models.
passing request.url as a string to the model is very similar to passing  species.image in the updated dictionary.
 The only change is that we move the data logic to the model.

Reply:
Model expects image as input. Note that the image would base_url + <species_taxonomy_id>.svg. Moreover image is string so I can not assign speciesTaxonomyId since it is number. I could convert it to str and then assign it and later in field_serialiser it will give the URL. 
So we have 2 options 
1) pass base_url, convert species_taxonomy_id to str and assign it to image , implement steriliser
2) just form the URL based on the species_taxonomy_id and pass it to model

Reply:
Alternatively, one more option is 

1. to introduce class variable base_url which will be set once/request.
2. Make image Nullable so while generating the PopularSpecies it won't have image
3. Implement serialised and for the URL based on the class variable base_url and species_taxonomy_id

Reply:
Serialisers are best when we want to modify the output of a model which would cause the issues that you just mentioned.
We can add a root_validator to access all the values before they are assigned to fields, for either PopularSpeciesGroup or PopularSpecies.
 For PopularSpeciesGroup to do similar to what you have done in the view would look like the following:
Model:
`class PopularSpeciesGroup(BaseModel):
    _request_url: str  # A hidden field to store request.url
    popular_species: List[PopularSpecies]

    @root_validator(pre=True)
    def generate_objects(cls, values):
        logger.info(values)
        for species in values["popular_species"]:
            species['image'] = '{}static/genome_images/{}.svg'.format(values["_request_url"],
                                                                      species['speciesTaxonomyId'])
        return values`

and the view would look like this: 
` popular_species_response = PopularSpeciesGroup(_request_url=request.base_url,popular_species=popular_species)`

We can do exactly the same in PopularSpecies and it wouldn't need the for loop anymore because data will be updated automatically before it's assigned to the fields.